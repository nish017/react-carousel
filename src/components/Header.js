import React from 'react'

// functional component
const Header = () => {

    // Below JSX renders the header bar
    return (
        <div className="carousel-header">
            <h1>Carousel Test</h1>
        </div> 
    )
}

export default Header