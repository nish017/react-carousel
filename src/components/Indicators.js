import React from 'react'

// functional component
const Indicator = (props) => {
 
    return (      
        <div className={`indicator ${(props.active >= props.index*props.total && 
            props.active < (props.index+1)*props.total) ? 'selected' : ''}`} 
            onClick={() => {props.slideToView(props.index, props.total)}}>
        </div>
    )
}

export default Indicator