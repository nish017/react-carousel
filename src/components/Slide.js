import React from 'react'

// functional component
const Slide = (props) => {

    const title = props.data.tags.split(',')[0]

    // Slide JSX renders individual slide of carousel, props- active and index helps achieve sliding functionality
    return (        
        <div className={`carousel-slide col-12 col-md-4 col-lg-2 ${props.active == props.index ? 'active' : ''}`}
        style={{'transform': `translateX(-${props.active*100}%) ${props.active == props.index ? 'scale(1.1)' : ''}`}}>
            <div>
                <img src={props.data.webformatURL} />
            </div>
            <h4 className="mt-2">{title}</h4>
        </div>
    )
}

export default Slide