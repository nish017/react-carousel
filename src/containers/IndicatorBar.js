import React from 'react'
import Indicator from '../components/Indicators'

class IndicatorBar extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            slidesPerView: 1
        }
    }

    componentDidMount () {
        if (window.screen.width > 992) {
            this.setState({
                slidesPerView: 5
            })
        } else if(window.screen.width > 768 && window.screen.width < 992) {
            this.setState({
                slidesPerView: 3
            })
        }
    }

    render() {
        var indicators = [];
        if(this.props.data) {
            for (var i = 0; i < Math.ceil(this.props.data.length/this.state.slidesPerView); i++) {
                indicators.push(<Indicator key={i} total={this.state.slidesPerView} 
                index={i} active={this.props.active} slideToView={this.props.slideToView}/>);
            }
        }
        
        return (
            <div className="row justify-content-center footer">
                {indicators}
            </div>
        )
    }
}

export default IndicatorBar