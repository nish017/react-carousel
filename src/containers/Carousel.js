import React, { Component } from 'react'
import Slide from '../components/Slide'

//class based react component
class Carousel extends Component {
    constructor(props) {
        super(props)
    }

    render() {
        return (
            <div>
                <div className="carousel">
                    <div className="carousel-inner">
                        {this.props.data.map((slideData) => <Slide key={slideData.id} data={slideData} />)}
                    </div>
                </div>
                <div className="carousel-footer">
                    <div className="row justify-content-center footer">
                        <button className="btn prev">Prev</button>
                        <button className="btn next">Next</button>
                    </div>
                </div>
                
            </div>
        )
    }
}

export default Carousel