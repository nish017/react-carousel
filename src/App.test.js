
import React from 'react';
import renderer from 'react-test-renderer';
import Slide from './components/Slide';
import Header from './components/Header';
import Carousel from './containers/Carousel';
import Indicators from './components/Indicators';
import IndicatorBar from './containers/IndicatorBar';

test('<Slide /> renders correctly', () => {
    const data = {
        tags: "a,b,c"
    }
  const tree = renderer
    .create(<Slide data={data} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('<Header /> renders correctly', () => {
  const tree = renderer
    .create(<Header />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('<Carousel /> renders correctly', () => {
  const tree = renderer
    .create(<Carousel />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('<Indicators /> renders correctly', () => {
  const tree = renderer
    .create(<Indicators />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('<IndicatorBar /> renders correctly', () => {
  const tree = renderer
    .create(<IndicatorBar />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('<Carousel /> component: showPrevSlide fn', () => {
    const carouselInstance = renderer.create(<Carousel />).getInstance();
    const e = { stopPropagation: jest.fn() };
    carouselInstance.showPrevSlide(e)
    expect(carouselInstance.state.active).toBe(-1)
});

test('<Carousel /> component: showNextSlide fn', () => {
    const carouselInstance = renderer.create(<Carousel />).getInstance();
    const e = { stopPropagation: jest.fn() };
    carouselInstance.showNextSlide(e)
    expect(carouselInstance.state.active).toBe(1)
});

test('<Carousel /> component: slideToView fn next views', () => {
  const carouselInstance = renderer.create(<Carousel />).getInstance();
  carouselInstance.slideToView(1,5)
  expect(carouselInstance.state.active).toBe(5)
});

test('<Carousel /> component: slideToView fn for prev view', () => {
  const carouselInstance = renderer.create(<Carousel />).getInstance();
  carouselInstance.slideToView(-1,5)
  expect(carouselInstance.state.active).toBe(-5)
});
