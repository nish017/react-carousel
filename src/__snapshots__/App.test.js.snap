// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel /> renders correctly 1`] = `
<div>
  <div
    className="carousel"
  >
    <div
      className="carousel-inner"
    />
    <img
      className="arrow-prev disabled-arrow"
      onClick={null}
      src="test-file-stub"
    />
    <img
      className="arrow-next "
      onClick={[Function]}
      src="test-file-stub"
    />
  </div>
  <div
    className="carousel-footer d-none d-md-block"
  >
    <div
      className="row justify-content-center footer"
    >
      <button
        className="btn prev"
        disabled="disabled"
        onClick={[Function]}
      >
        Prev
      </button>
      <button
        className="btn next"
        disabled=""
        onClick={[Function]}
      >
        Next
      </button>
    </div>
  </div>
</div>
`;

exports[`<Header /> renders correctly 1`] = `
<div
  className="carousel-header"
>
  <h1>
    Carousel Test
  </h1>
</div>
`;

exports[`<Slide /> renders correctly 1`] = `
<div
  className="carousel-slide col-12 col-md-4 col-lg-2 active"
  style={
    Object {
      "transform": "translateX(-NaN%) scale(1.1)",
    }
  }
>
  <div>
    <img />
  </div>
  <h4
    className="mt-2"
  >
    a
  </h4>
</div>
`;

exports[`renders correctly 1`] = `
<div
  className="carousel-slide col-12 col-md-4 col-lg-2 active"
  style={
    Object {
      "transform": "translateX(-NaN%) scale(1.1)",
    }
  }
>
  <div>
    <img />
  </div>
  <h4
    className="mt-2"
  >
    a
  </h4>
</div>
`;
